### PR TITLE
Fix Issue 21058 - __traits(getOverloads) forgets "this" with third ar…

### PR DESCRIPTION
**Pull request by rmanthorpe**
…gument or if first overload is a template

### CODE REVIEW

**Comment by UplinkCoder** [comment]
can you do it without changing the name `e` to `exp` ?

**Comment by rmanthorpe** [comment]
The local `e` was shadowing another `e`. Is it a problem? I can just rearrange to put more into the `if(fd)` block.

File: src/dmd/traits.d [hunk 1]
```diff
@@ -1047,9 +1047,11 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
                 f = td;
                 if (td && td.funcroot)
                     f = td.funcroot;
-                ex = null;
+                if (dte.e1.op == TOK.dotType || dte.e1.op == TOK.this_)
```
Comment (UplinkCoder) [inline src/dmd/traits.d:1050]: I would negate this branch and leave the default ex = null.

**Review by UplinkCoder — APPROVED** [review]
Looks good to me.
Except for one comment I made.

**Comment by UplinkCoder** [comment]
very nice and small now

**Review by ghost — APPROVED** [review]

File: src/dmd/traits.d [hunk 2]
```diff
@@ -1076,14 +1077,12 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
             int dg(Dsymbol s)
             {
-                if (includeTemplates)
-                {
-                    exps.push(new DsymbolExp(Loc.initial, s, false));
-                    return 0;
-                }
                 auto fd = s.isFuncDeclaration();
-                if (!fd)
+                if (!fd) {
```
Comment (NilsLankila) [inline src/dmd/traits.d:1081]: Left curly brace on its own line, [as specified in point 3](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#the-following-will-not-be-viewed-with-favor).

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Comment by NilsLankila** [comment]
Thanks forf this bugfix @rmanthorpe, just one thing is that for small fixups to the main commit you should rather amend, but the "auto-merge-squash" label should do more or less the same.

**Comment by rmanthorpe** [comment]
Thanks @NilsLankila I'll bear that in mind in the future!

**Pull request merged**